LS240 Locking restructure

jlashner opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR provides an example of how agents can be re-written to avoid locking mechanisms, using the single `main` process discussed in daq-dev


## Description
<!--- Describe your changes in detail -->
Refactors LS240 agent to use lock-free agent structure as discussed in daq-dev. This is not _really_ ready for PR, but I think it is far enough for people to look at / we can discuss structural changes or what we may want to move over to ocs proper.

A few notes about how it works:
 - Actions are defined as dataclasses with a `process` function that takes in some hardware driver object and does things with it.
 - tasks that create new actions will yield until the action is processed, at which point, anything returned from `process` will be set to `session.data` of said task. If the `process` function raises an exception, the defered errback is called, in which case the `yield`  statement for that action will throw an exception, so the task will crash, while the `main` process will catch the exception and keep going.
 - I opted to define a `process` function per action, instead of a single `process` function that takes in an action type, because I think that will be easier if we move boilerplate into the ocs repo, but I'm not super attached
 -  Error handling / robustness needs to be checked with real hardware, I'm not entirely sure I have the correct exceptions for when the module loses connection since I've only tested with a simulator
 - Still need to add docs
 - "degraded" process status / feed info not included

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Removes multithreading / locking requirements from hardware handling in ls240

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Tested using ls240 simulator

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

jlashner commented:
It's been a while since I started this, but I'd like to open this as a real PR since I think it would really improve the agent, and "robustifies" the agent as is described here: https://github.com/simonsobs/socs/discussions/538

In addition, I added a bunch of types and statically checked this module with mypy. I really liked this and it was useful for catching a number of cases in which optional types were not handled correctly, or the correct values weren't being returned from OCS operations. I think we should consider adding types and running OCS core through mypy, and encourage new agents/developers to use it.

BrianJKoopman commented on socs/agents/lakeshore240/agent.py:
```diff
@@ -1,82 +1,139 @@
-                                 agg_params=agg_params,
-                                 buffer_time=1)
+        def __post_init__(self) -> None:
+            self.processed: bool = False
+            self.success: bool = False
+            self.traceback: Optional[str] = None
+            self.result: Any = None
```
If this structure ends up sticking around these should be private, else agent authors need to be sure not to collide any agent arguments with these (or any future) attributes within the `BaseAction` -- I imagine once this is more solidified `BaseAction` and similar machinery would move to ocs-core.

BrianJKoopman reviewed (commented):
Thanks for working on this restructure! And sorry this took so long to review. I'm excited about some of the things this improves, like making error handling more straightforward, since it isolates communication with the hardware within the `main` process.

A lot of what I wrote is my opinion on the overall structure from the perspective of if we were to make this the new recommended way that contributors write agents. That said, since we have several other agents converted to this structure already, and since this likely solves a few of the open LS240 agent issues, I'd be fine merging this with minimal changes. I just wanted to take the opportunity to discuss the structure since that's how this PR was initially framed.

jlashner commented on tests/integration/test_ls240_agent_integration.py:
```diff
@@ -7,112 +7,81 @@
-                             units=1,
-                             enabled=1,
-                             name="Channel 01")
+    set_values_params = LS240Actions.SetValues(
+        channel=1, sensor=1, auto_range=1, range=0, current_reversal=0, units=1,
+        enabled=1, name="Channel 01"
+    )
+    resp = client.set_values(**asdict(set_values_params))
```
Clients don't need to work this way, and using regular keyword args still works, however I think it offers a couple of benefits... Mainly it ensures that client keyword args are what the agent expects, and it allows for static type checking on the client-side. If there is an unexpected parameter, a missing required parameter, or a type mismatch it can be caught here statically as opposed to requiring an integration test.

BrianJKoopman commented on socs/agents/lakeshore240/agent.py:
```diff
@@ -1,82 +1,122 @@
-                self.log.warn("Could not start init because "
-                              "{} is already running".format(self.lock.job))
-                return False, "Could not acquire lock."
+    @dataclass
+    class SetValues(BaseAction):
+        """
+        Action to configure settings on the lakeshore.  For parameter details,
+        see the docstrings for the `set_values` task.
+        """
```
> Maybe I'm missing something, but I don't really see how the actions can be functions, since they have to actually store the data that tells the `main` process what parameters to use when processing the action.

Instead of storing the params as class attributes and putting the class in the queue you could put both the function and params in the queue, then use them in `_process_actions` -- kind of like how [`AgentOp` launches functions](https://github.com/simonsobs/ocs/blob/fe0342163e8a2480545df6015eddcfc86b97079b/ocs/ocs_agent.py#L982-L984). An example when I was thinking about this before the `Deferred` was dropped:
```python
def _process_actions(self, module: Module) -> None:
    """
    Processes queued actions using the provided Lakeshore Module.
    """
    while not self.action_queue.empty():
        # ActionClass(**params) -> (function, params, deferred)
        task, params, d_ = self.action_queue.get()
        try:
            log.info(f"Running action {task}")
            res = task(module, **params)
            d_.callback(res)
        except Exception as e:  # pylint: disable=broad-except
            log.error(f"Error processing action: {task}")
            d_.errback(e)
    return None
```

> The way I imagined solving this duplication of code is eliminating the task definition entirely (i.e. the `set_values` function in the LS240_Agent class), since this is almost entirely identical across different actions and is mostly boilerplate. Instead, we could have a `register_task_from_action` function, that took in the Action class and generated the OCS endpoint from it. Then we'd only have to document the parameters once, in the class definition.

Ah, that wasn't clear from the PR. This still seems like it would present as a class docstring, when we want to present a method docstring to users, since on the client end `process()` is hidden from them.

Maybe my above suggestion isn't the best either, but the launching machinery will be hidden from developers if this gets moved into a `register_task_from_action` function. I'm mostly concerned with the developer/user experience on the agent end.

jlashner commented on socs/agents/lakeshore240/agent.py:
```diff
@@ -108,202 +148,211 @@ def set_values(self, session, params=None):
+                action.deferred.errback(e)
+        return None
+
+    def main(
```
That's a good question, I think there isn't a clear answer how multiple processes should work. In general, multiple processes like that in general kind of breaks the hardware "single point of access" I'm going for here. I think it would depend on a case to case basis. Like you say, for the custom_pid function would probably be an argument you could pass into the main function to swap out the behavior.

jlashner commented:
@BrianJKoopman I just added a function to automatically register tasks (with proper ocs_params) based on the Action class. I tested with integration tests, and it works as expected (fails automatically for parameter mismatch, etc.). Do you think you can take another look and see if this seems reasonable to you and solves your concerns about duplication?

BrianJKoopman commented on socs/util.py:
I'm not a fan of `util.py`, it'll just accrues a bunch of miscellaneous stuff. Can we put the new actions stuff in a module called `action.py`? I imagine this'll move upstream to `ocs` eventually. (`get_md5sum` only gets imported by one module, we should frankly move it and get rid of `util.py`, but not in this PR.)

BrianJKoopman commented on socs/util.py:
```diff
@@ -1,9 +1,195 @@
+    """
+    try:
+        isinstance(0, t)
+        return True
+    except Exception:
+        return False
+
+
+def get_param_type(t: Type) -> Optional[Type]:
+    """
+    OCS param type variables require you to be able to run isinstance,
+    which does not work for subscripted generics like Optional in python3.8.
+    This function attempts to convert types to values that will be accepted
+    by ocs_agent.param, unwrapping optional types if we are unable to run
+    isinstance off the bat.
+
+    Other subscripted generics such as List[...] or Dict[...] are not currently
+    supported.
+
+    This function will return the unwrapped type, or None if it fails.
+    """
+    origin_type = get_origin(t)
+
+    # Unwrap possible option type
+    if origin_type == Union:
+        sub_types = get_args(t)
+        if len(sub_types) != 2:
+            return None
+        if type(None) not in sub_types:
+            return None
+        for st in sub_types:
+            if st is not type(None):
+                if is_instanceable(st):
+                    return st
+
+    elif is_instanceable(t):
+        # If this works, then it should work with ocs_agent.param
+        return t
+
+    return None
```
Let's mark these two functions as private with a leading `_`. They can always be made public later if needed.

BrianJKoopman reviewed (changes requested):
> @BrianJKoopman I just added a function to automatically register tasks (with proper ocs_params) based on the Action class. I tested with integration tests, and it works as expected (fails automatically for parameter mismatch, etc.). Do you think you can take another look and see if this seems reasonable to you and solves your concerns about duplication?

Great, `register_task_from_action()` looks good! Reducing that boilerplate is definitely moving in the right direction. Comments below. If you could give some thoughts on the remaining comments from the initial review too that'd be great.

jlashner commented on socs/util.py:
```diff
@@ -1,9 +1,195 @@
+
+    task.__doc__ = action_class.__doc__
+
+    # Adds ocs parameters
+    for f in dataclasses.fields(action_class):
+        param_type = get_param_type(f.type)
+        if param_type is None:
+            raise ValueError(f"Unsupported param type for arg {f.name}: {f.type}")
+        param_kwargs: Dict[str, Any] = {
+            "type": param_type,
+        }
+        if f.default != dataclasses.MISSING:
+            param_kwargs["default"] = f.default
+        if isinstance(f.metadata, dict):
+            param_kwargs.update(f.metadata.get("ocs_param_kwargs", {}))
+        task = ocs_agent.param(f.name, **param_kwargs)(task)
```
Yes! additional kwargs can be specified by setting the `ocs_param_kwargs`  key in the dataclass field metadata, for instance:
```
from dataclases import dataclass, field
...
@dataclass
class ExampleAction(LS240Action):
    current: float = field(
        metadata={'ocs_param_kwargs': {'check': lambda x: 0.0 <= x <= 2.0}}
    )
```
A bit verbose, but provides this functionality for those who needs it.

jlashner commented on socs/agents/lakeshore240/agent.py:
```diff
@@ -1,309 +1,326 @@
+            DeprecationWarning,
+        )
 
     if args.num_channels is not None:
-        warnings.warn("WARNING: the --num-channels parameter is deprecated, please "
-                      "remove from your site-config file", DeprecationWarning)
-
-    # Automatically acquire data if requested (default)
-    init_params = False
-    if args.mode == 'init':
-        init_params = {'auto_acquire': False}
-    elif args.mode == 'acq':
-        init_params = {'auto_acquire': True}
+        warnings.warn(
+            "WARNING: the --num-channels parameter is deprecated, please "
+            "remove from your site-config file",
+            DeprecationWarning,
+        )
+
+    if args.mode is not None:
+        warnings.warn(
+            "WARNING: the --init-mode parameter is deprecated, please "
+            "remove from your site-config file",
+            DeprecationWarning,
+        )
 
     device_port = None
     if args.port is not None:
         device_port = args.port
     else:  # Tries to find correct USB port automatically
-
         # This exists if udev rules are setup properly for the 240s
-        if os.path.exists('/dev/{}'.format(args.serial_number)):
+        if os.path.exists("/dev/{}".format(args.serial_number)):
             device_port = "/dev/{}".format(args.serial_number)
 
-        elif os.path.exists('/dev/serial/by-id'):
-            ports = os.listdir('/dev/serial/by-id')
+        elif os.path.exists("/dev/serial/by-id"):
+            ports = os.listdir("/dev/serial/by-id")
```
Yes... sorry I did this in the PR, but I did run this through Ruff auto-formatting.  Its very nice and I think would be a good addition to the pre-commit. Since this is a big rewrite I thought it would be ok to run on this file, but in general it can distract from actual changes, so I think we'd want to run on the whole repo before implementing the pre-commit.

jlashner commented on socs/agents/lakeshore240/agent.py:
```diff
@@ -1,309 +1,326 @@
+
+@dataclass
+class UploadCalCurve(LS240Action):
+    """upload_cal_curve(channel, filename)
+
+    **Task** - Upload a calibration curve to a channel.
+
+    Args
+    ------
+    channel (int):
+        Channel number, 1-8.
+    filename (str):
+        Filename for calibration curve.
+    """
+
```
Would it be ok if we adopted different docstring standards for this type of class? I.e. just document the class regularly.

BrianJKoopman replied:
I don't think so. The user of the documentation is most commonly someone who wants to interact on the client side, which is why I think it should present like a function. Those users don't need to know about the `process()` method at all (in this example).

jlashner commented:
Hi Brian,

I went back through and addressed your comments, so if you can give it another look that would be great!

After thinking about the documentation issue, I think that the awkwardness stems from the fact that we are using docs of agent class members and supporting elements as the documentation of the client interface, even when the two don't match perfectly.  I think this documentation method enforces pretty strict structural requirements for the agent class, which isn't always necessary or beneficial.

While using the agent method docstrings as the official "agent API" is nice for a lot of agents, I think we should also support agents that are structured differently, and figure out another way to document the  API for these cases, even if its just writing it up manually in the docs page.

For now, replaced the `LS240Agent` autoclass in the docs page with the docs for the Action classes, and in the action class docstrings I put examples for how to run the action through a client. Hopefully this strikes a balance.

Thanks!
